src/libraries/Microsoft.Bcl.TimeProvider/src/System/Threading/Tasks/TimeProviderTaskExtensions.cs

```diff
@@ -0,0 +1,171 @@
+        {
+            if (timeProvider == TimeProvider.System)
+            {
+                return Task.Delay(delay, cancellationToken);
+            }
+
+            if (timeProvider is null)
+            {
+                throw new ArgumentNullException(nameof(timeProvider));
+            }
+
+            if (delay != Timeout.InfiniteTimeSpan && delay < TimeSpan.Zero)
+            {
+                throw new ArgumentOutOfRangeException(nameof(delay));
+            }
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return Task.FromCanceled(cancellationToken);
+            }
+
+            if (delay == TimeSpan.Zero)
+            {
+                return Task.CompletedTask;
+            }
+
+            DelayState state = new();
+            state.Timer = timeProvider.CreateTimer(delayState =>
+            {
+                DelayState s = (DelayState)delayState;
+                s.TrySetResult(true);
+                s.Registration.Dispose();
+                s.Timer.Dispose();
+            }, state, Timeout.InfiniteTimeSpan, Timeout.InfiniteTimeSpan);
+
+            state.Timer.Change(delay, Timeout.InfiniteTimeSpan);
+
+            state.Registration = cancellationToken.Register(delayState =>
+            {
+                DelayState s = (DelayState)delayState;
+                s.TrySetCanceled(cancellationToken);
+                s.Timer.Dispose();
+                s.Registration.Dispose();
+            }, state);
+
+            if (state.Task.IsCompleted)
+            {
+                state.Registration.Dispose();
+            }
+
+            return state.Task;
+        }
+
+        private sealed class WaitAsyncState : TaskCompletionSource<bool>
+        {
+            public WaitAsyncState() : base(TaskCreationOptions.RunContinuationsAsynchronously) { }
+            public readonly CancellationTokenSource ContinuationCancellation = new CancellationTokenSource();
+            public CancellationTokenRegistration Registration;
+            public ITimer? Timer;
+        }
+
+        public static Task WaitAsync(this Task task, TimeSpan timeout, TimeProvider timeProvider, CancellationToken cancellationToken = default)
+        {
+            if (task is null)
+            {
+                throw new ArgumentNullException(nameof(task));
+            }
+
+            if (timeout != Timeout.InfiniteTimeSpan && timeout < TimeSpan.Zero)
+            {
+                throw new ArgumentOutOfRangeException(nameof(timeout));
+            }
+
+            if (timeProvider is null)
+            {
+                throw new ArgumentNullException(nameof(timeProvider));
+            }
+
+            if (task.IsCompleted)
+            {
+                return task;
+            }
+
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return Task.FromCanceled(cancellationToken);
+            }
```

Comment:
You probably similarly want a check for timeout == TimeSpan.Zero, immediately returning a timed out task.